service/controller/v17patch1/resource/endpoint/current.go

```diff
@@ -85,7 +85,8 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 		r.logger.LogCtx(ctx, "level", "debug", "message", "found endpoint")
 	}
 
-	if key.IsPodDeleted(pod) {
+	// Remove master endpoint ip only after pod is fully terminated because we still need to drain master before we remove it.
+	if serviceName == key.MasterID && key.IsPodDeleted(pod) {
```

Comment:
Are you sure you want to check for the master here? I feel like this change has undesired side effects and you actually want to add the check below in combination with `ArePodContainersTerminated`.

Reply:
Whole code below `key.IsPodDeleted(pod)` does not make sense for `worker` endpoint.

For the worker, we want to remove the IP from endpoint immediately. And the code below just fetches state of pod and checks if it's in the terminated state or not. If the containers are in the terminated state that means they no longer can accept traffic and we would still sending traffic to it until that point which results in downtime.


```
currentPod, err := r.k8sClient.CoreV1().Pods(pod.GetNamespace()).Get(pod.GetName(), metav1.GetOptions{})
```
is there only so we have the latest version of the pod and the rest is error handling of this command and checking for the terminated state.

This is intented

Reply:
Why do we want to terminate for workers right away? Why not wait for terminated state?

Reply:
We don't terminate workers, we remove IP from endpointList.

The reason why is that if we wait until the container terminated we will lose traffic.

Reply:
this change si about endpoint resource only